# Tucil IF3270 Pembelajaran Mesin

Naufal Alexander Suryasumirat - 13519135

Naufal Yahya Kurnianto - 13519141

## Import Libraries

In [1]:
import six
import sys
sys.modules['sklearn.externals.six'] = six

In [2]:
import id3
import sklearn
import tabulate
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.svm import SVC
from sklearn import compose
from sklearn import preprocessing
from sklearn import tree, datasets
from sklearn.cluster import KMeans
from sklearn.tree import export_text
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, accuracy_score
from sklearn.model_selection import train_test_split
from id3 import Id3Estimator, export_graphviz, export_text

In [3]:
# Function to print model parameters
def show_param(model, custom_string):
  print(custom_string)
  params = model.get_params().items()
  for i in params:
    print(f"{i[0]}: {i[1]}")

## Soal 1 - Membaca Dataset (Load Dataset)

### Load Dataset Breast Cancer

In [4]:
# Load Dataset Breast Cancer
data_bc = sklearn.datasets.load_breast_cancer()
# Splitting Data Sets (80% training, 20% scoring/validating/testing)
x, y = data_bc.data, data_bc.target
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 420)
y_true = y_test

### Load Dataset Play-Tennis

In [5]:
url = 'https://gist.githubusercontent.com/DiogoRibeiro7/c6590d0cf119e87c39e31c21a9c0f3a8/raw/4a8e3da267a0c1f0d650901d8295a5153bde8b21/PlayTennis.csv'

In [6]:
# Load Dataset Play-Tennis
data_pt = pd.read_csv(url)
x = data_pt.loc[:, ['Outlook', 'Temperature', 'Humidity', 'Wind']]
y = data_pt['Play Tennis']
x = x.apply(preprocessing.LabelEncoder().fit_transform)
# Using LabelEncoder to convert Yes -> 1 and No -> 0
le = preprocessing.LabelEncoder()
y = le.fit_transform(y)
# Splitting Data Sets (80% training, 20% scoring/validating/testing)
x_train_pt, x_test_pt, y_train_pt, y_test_pt = train_test_split(x, y, test_size = 0.2, random_state = 1)

y_true_pt = y_test_pt

## Soal 2a & 3a - Pembelajaran, Prediksi, dan Evaluasi dengan Algoritma DecisionTreeClassifier

### Breast Cancer Dataset

In [7]:
# DecisionTreeClassifier (Breast Cancer)
clf_dtc_bc = sklearn.tree.DecisionTreeClassifier().fit(x_train, y_train)
show_param(clf_dtc_bc, "Parameters for DecisionTreeClassifier model Dataset Breast Cancer:")

Parameters for DecisionTreeClassifier model Dataset Breast Cancer:
ccp_alpha: 0.0
class_weight: None
criterion: gini
max_depth: None
max_features: None
max_leaf_nodes: None
min_impurity_decrease: 0.0
min_impurity_split: None
min_samples_leaf: 1
min_samples_split: 2
min_weight_fraction_leaf: 0.0
random_state: None
splitter: best


In [8]:
# Export DecicisionTreeClassifier as Text (Breast Cancer)
print(tree.export_text(clf_dtc_bc, feature_names = data_bc.feature_names.tolist()))

|--- worst perimeter <= 114.45
|   |--- worst concave points <= 0.16
|   |   |--- area error <= 33.00
|   |   |   |--- worst concave points <= 0.13
|   |   |   |   |--- worst texture <= 33.27
|   |   |   |   |   |--- class: 1
|   |   |   |   |--- worst texture >  33.27
|   |   |   |   |   |--- worst texture <= 33.56
|   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |--- worst texture >  33.56
|   |   |   |   |   |   |--- class: 1
|   |   |   |--- worst concave points >  0.13
|   |   |   |   |--- worst texture <= 29.06
|   |   |   |   |   |--- concave points error <= 0.01
|   |   |   |   |   |   |--- mean smoothness <= 0.10
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- mean smoothness >  0.10
|   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- concave points error >  0.01
|   |   |   |   |   |   |--- class: 1
|   |   |   |   |--- worst texture >  29.06
|   |   |   |   |   |--- class: 0
|   |   |--- area error >  33.00
|   |   |   |--- wors

In [9]:
# Predicting using DecisionTreeClassifer for Breast Cancer
y_pred_dtc_bc = clf_dtc_bc.predict(x_test)

In [10]:
# Scoring DecisionTreeClassifier for Breast Cancer (Accuracy)
acc_dtc_bc = accuracy_score(y_true, y_pred_dtc_bc)
print("Accuracy Score DecisionTreeClassifier (Dataset Breast Cancer) =", acc_dtc_bc)

Accuracy Score DecisionTreeClassifier (Dataset Breast Cancer) = 0.9298245614035088


In [11]:
# Scoring DecisionTreeClassifier for Breat Cancer (F1)
f1_dtc_bc = f1_score(y_true, y_pred_dtc_bc, pos_label = 1)
print("F1 Score DecisionTreeClassifier (Dataset Breast Cancer) =", f1_dtc_bc)

F1 Score DecisionTreeClassifier (Dataset Breast Cancer) = 0.9420289855072465


### Play-Tennis Dataset

In [12]:
# DecisionTreeClassifier (Play-Tennis)
clf_dtc_pt = sklearn.tree.DecisionTreeClassifier().fit(x_train_pt, y_train_pt)
show_param(clf_dtc_pt, "Parameters for DecisionTreeClassifier model Dataset Play-Tennis:")

Parameters for DecisionTreeClassifier model Dataset Play-Tennis:
ccp_alpha: 0.0
class_weight: None
criterion: gini
max_depth: None
max_features: None
max_leaf_nodes: None
min_impurity_decrease: 0.0
min_impurity_split: None
min_samples_leaf: 1
min_samples_split: 2
min_weight_fraction_leaf: 0.0
random_state: None
splitter: best


In [13]:
# Export DecicisionTreeClassifier as Text (Play-Tennis)
print(tree.export_text(clf_dtc_pt, feature_names = data_pt.columns.values[:-1].tolist()))

|--- Outlook <= 0.50
|   |--- class: 1
|--- Outlook >  0.50
|   |--- Humidity <= 0.50
|   |   |--- class: 0
|   |--- Humidity >  0.50
|   |   |--- Wind <= 0.50
|   |   |   |--- Temperature <= 1.00
|   |   |   |   |--- class: 0
|   |   |   |--- Temperature >  1.00
|   |   |   |   |--- class: 1
|   |   |--- Wind >  0.50
|   |   |   |--- class: 1



In [14]:
# Predicting using DecisionTreeClassifer for Play-Tennis
y_pred_dtc_pt = clf_dtc_pt.predict(x_test_pt)

In [15]:
# Scoring DecisionTreeClassifier for Play-Tennis (Accuracy)
acc_dtc_pt = accuracy_score(y_true_pt, y_pred_dtc_pt)
print("Accuracy Score DecisionTreeClassifier (Dataset Play-Tennis) =", acc_dtc_pt)

Accuracy Score DecisionTreeClassifier (Dataset Play-Tennis) = 0.6666666666666666


In [16]:
# Scoring DecisionTreeClassifier for Play-Tennis (F1)
f1_dtc_pt = f1_score(y_true_pt, y_pred_dtc_pt, pos_label = 1)
print("F1 Score DecisionTreeClassifier (Dataset Play-Tennis) =", f1_dtc_pt)

F1 Score DecisionTreeClassifier (Dataset Play-Tennis) = 0.6666666666666666


## Soal 2b & 3b - Pembelajaran, Prediksi, dan Evaluasi dengan Algoritma Id3Estimator

### Breast Cancer Dataset

In [17]:
# Id3Estimator (Breast Cancer)
estimator = Id3Estimator().fit(x_train, y_train)
show_param(estimator, "Parameters for Id3Estimator model Dataset Breast Cancer:")

Parameters for Id3Estimator model Dataset Breast Cancer:
gain_ratio: False
is_repeating: False
max_depth: None
min_entropy_decrease: 0.0
min_samples_split: 2
prune: False


In [18]:
# Export as Text Id3Estimator (Breast Cancer)
print(id3.export_text(estimator.tree_, data_bc.feature_names.tolist()))


worst perimeter <=117.45
|   worst concave points <=0.11
|   |   area error <=45.17
|   |   |   worst texture <=33.35: 1 (229) 
|   |   |   worst texture >33.35
|   |   |   |   mean texture <=23.84
|   |   |   |   |   mean radius <=12.07: 1 (1) 
|   |   |   |   |   mean radius >12.07: 0 (1) 
|   |   |   |   mean texture >23.84: 1 (15) 
|   |   area error >45.17
|   |   |   worst texture <=22.76: 1 (4) 
|   |   |   worst texture >22.76: 0 (4) 
|   worst concave points >0.11
|   |   worst texture <=25.83
|   |   |   mean smoothness <=0.13
|   |   |   |   worst area <=810.30: 1 (21) 
|   |   |   |   worst area >810.30
|   |   |   |   |   mean radius <=14.19: 0 (2) 
|   |   |   |   |   mean radius >14.19: 1 (8) 
|   |   |   mean smoothness >0.13: 0 (2) 
|   |   worst texture >25.83
|   |   |   mean concave points <=0.04
|   |   |   |   mean compactness <=0.09: 0 (2) 
|   |   |   |   mean compactness >0.09: 1 (9) 
|   |   |   mean concave points >0.04
|   |   |   |   mean concavity <=0.09


In [19]:
# Predicting using Id3Estimator for Breast Cancer
y_pred_id3_bc = estimator.predict(x_test)

In [20]:
# Scoring Id3Estimator for Breast Cancer (Accuracy)
acc_id3_bc = accuracy_score(y_true, y_pred_id3_bc)
print("Accuracy Score Id3Estimator (Dataset Breast Cancer) =", acc_id3_bc)

Accuracy Score Id3Estimator (Dataset Breast Cancer) = 0.9210526315789473


In [21]:
# Scoring Id3Estimator for Breast Cancer (F1)
f1_id3_bc = f1_score(y_true, y_pred_id3_bc, pos_label = 1)
print("F1 Score Id3Estimator (Dataset Breast Cancer) =", f1_id3_bc)

F1 Score Id3Estimator (Dataset Breast Cancer) = 0.9323308270676691


### Play-Tennis Dataset

In [22]:
# Id3Estimator (Play Tennis)
pt_estimator = Id3Estimator().fit(x_train_pt,y_train_pt)
show_param(pt_estimator, "Parameters for Id3Estimator model Dataset Play Tennis:")

Parameters for Id3Estimator model Dataset Play Tennis:
gain_ratio: False
is_repeating: False
max_depth: None
min_entropy_decrease: 0.0
min_samples_split: 2
prune: False


In [23]:
# Export as Text Id3Estimator (Play Tennis)
print(id3.export_text(pt_estimator.tree_, data_pt.columns.values[:-1].tolist()))


Outlook <=0.50: 1 (3) 
Outlook >0.50
|   Humidity <=0.50: 0 (3) 
|   Humidity >0.50
|   |   Wind <=0.50
|   |   |   Temperature <=1.00: 0 (1) 
|   |   |   Temperature >1.00: 1 (1) 
|   |   Wind >0.50: 1 (3) 



In [24]:
# Predicting using Id3Estimator for Play Tennis
y_pred_id3_pt = pt_estimator.predict(x_test_pt)

In [25]:
# Scoring Id3Estimator for Play Tennis (Accuracy)
acc_id3_pt = accuracy_score(y_true_pt, y_pred_id3_pt)
print("Accuracy Score Id3Estimator (Dataset Play Tennis) =", acc_id3_pt)

Accuracy Score Id3Estimator (Dataset Play Tennis) = 0.6666666666666666


In [26]:
# Scoring Id3Estimator for Play Tennis (F1)
f1_id3_pt = f1_score(y_true_pt, y_pred_id3_pt, pos_label = 1)
print("F1 Score Id3Estimator (Dataset Play Tennis) =", f1_id3_pt)

F1 Score Id3Estimator (Dataset Play Tennis) = 0.6666666666666666


## Soal 2c & 3c - Pembelajaran, Prediksi, dan Evaluasi dengan Algoritma K Means

### Breast Cancer Dataset

In [27]:
# KMeans (Breast Cancer) {Showing parameters}
kmeans = KMeans(n_clusters = 2, n_init = 15, max_iter = 600).fit(x_train)
show_param(kmeans, "Parameters for KMeans model Dataset Breast Cancer:")

Parameters for KMeans model Dataset Breast Cancer:
algorithm: auto
copy_x: True
init: k-means++
max_iter: 600
n_clusters: 2
n_init: 15
n_jobs: deprecated
precompute_distances: deprecated
random_state: None
tol: 0.0001
verbose: 0


In [28]:
# Showing Cluster center coordinates (Koordinat pusat centroid) for Breast Cancer
kmeans.cluster_centers_

array([[1.90807000e+01, 2.19466000e+01, 1.26249000e+02, 1.14803500e+03,
        1.02059900e-01, 1.48691500e-01, 1.74846500e-01, 9.95550000e-02,
        1.93514000e-01, 6.08909000e-02, 7.22024000e-01, 1.21749900e+00,
        5.05981000e+00, 9.13457000e+01, 6.65091000e-03, 3.19766000e-02,
        4.20957000e-02, 1.55785100e-02, 1.99333000e-02, 3.96172000e-03,
        2.34274000e+01, 2.93130000e+01, 1.56562000e+02, 1.71322000e+03,
        1.41873200e-01, 3.58207000e-01, 4.51690000e-01, 1.92432600e-01,
        3.14465000e-01, 8.71631000e-02],
       [1.25132648e+01, 1.84550423e+01, 8.08738592e+01, 4.92789014e+02,
        9.51919155e-02, 9.24783944e-02, 6.39625344e-02, 3.39360225e-02,
        1.78788732e-01, 6.37034648e-02, 3.07884789e-01, 1.21394113e+00,
        2.17172310e+00, 2.40197296e+01, 7.26076620e-03, 2.38344056e-02,
        2.91559200e-02, 1.06871775e-02, 2.07097972e-02, 3.81023521e-03,
        1.39989944e+01, 2.45205070e+01, 9.16678873e+01, 6.15860000e+02,
        1.30720225e-01,

In [29]:
# Predicting using K Means for Breast Cancer
y_pred_km_bc = kmeans.predict(x_test)

In [30]:
# Scoring KMeans for Breast Cancer (Accuracy)
acc_km_bc = accuracy_score(y_true, y_pred_km_bc)
print("Accuracy Score K Means (Dataset Breast Cancer) =", acc_km_bc)

Accuracy Score K Means (Dataset Breast Cancer) = 0.868421052631579


In [31]:
# Scoring KMeans for Breast Cancer (F1)
f1_km_bc = f1_score(y_true, y_pred_km_bc, pos_label = 1)
print("F1 Score K Means (Dataset Breast Cancer) =", f1_km_bc)

F1 Score K Means (Dataset Breast Cancer) = 0.8993288590604026


### Play-Tennis Dataset

In [32]:
# KMeans (Play Tennis)
pt_kmeans = KMeans(n_clusters = 2).fit(x_train_pt)
show_param(pt_kmeans, "Parameters for KMeans model Dataset Play Tennis:")

Parameters for KMeans model Dataset Play Tennis:
algorithm: auto
copy_x: True
init: k-means++
max_iter: 300
n_clusters: 2
n_init: 10
n_jobs: deprecated
precompute_distances: deprecated
random_state: None
tol: 0.0001
verbose: 0


In [33]:
# Showing Cluster center coordinates (Koordinat pusat centroid) for Play Tennis
pt_kmeans.cluster_centers_

array([[1.66666667, 0.66666667, 0.66666667, 0.5       ],
       [0.4       , 1.6       , 0.4       , 0.6       ]])

In [34]:
# Predicting using K Means for Play Tennis
y_pred_km_pt = pt_kmeans.predict(x_test_pt)

In [35]:
# Scoring KMeans for Play Tennis (Accuracy)
acc_km_pt = accuracy_score(y_true_pt, y_pred_km_pt)
print("Accuracy Score K Means (Dataset Play Tennis) =", acc_km_pt)

Accuracy Score K Means (Dataset Play Tennis) = 1.0


In [36]:
# Scoring KMeans for Play Tennis (F1)
f1_km_pt = f1_score(y_true_pt, y_pred_km_pt, pos_label = 1)
print("F1 Score K Means (Dataset Play Tennis) =", f1_km_pt)

F1 Score K Means (Dataset Play Tennis) = 1.0


## Soal 2d & 3d - Pembelajaran, Prediksi, dan Evaluasi dengan Algoritma LogisticRegression

### Breast Cancer Dataset

In [37]:
# LogisticRegression (Breast Cancer)
clf_lr_bc = LogisticRegression(max_iter = 1000).fit(x_train, y_train)
show_param(clf_lr_bc, "Parameters for LogisticRegression model Dataset Breast Cancer:")

Parameters for LogisticRegression model Dataset Breast Cancer:
C: 1.0
class_weight: None
dual: False
fit_intercept: True
intercept_scaling: 1
l1_ratio: None
max_iter: 1000
multi_class: auto
n_jobs: None
penalty: l2
random_state: None
solver: lbfgs
tol: 0.0001
verbose: 0
warm_start: False


C:\Users\Noler\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [38]:
# Showing Coefficient and Bias from Logistic Regression in Decision Function for Breast Cancer
print(clf_lr_bc.coef_,clf_lr_bc.intercept_)

[[ 1.57706927e+00  2.65894460e-01  1.50246140e-01 -1.28532970e-02
  -2.45862619e-01 -2.07516083e-01 -5.64412323e-01 -4.44045172e-01
  -3.03340951e-01  1.18334755e-02 -1.42877090e-04  3.73413688e-01
   4.28529674e-01 -1.36182196e-01 -1.34537737e-02  2.80334807e-01
   2.36596482e-01 -2.86860392e-02  1.71718592e-03  5.09703445e-02
   1.01213980e+00 -4.01148687e-01 -3.29487406e-01 -9.11919909e-03
  -4.69237311e-01 -4.58748610e-01 -1.35089490e+00 -8.97039445e-01
  -8.23448246e-01 -5.94355600e-02]] [7.58046412]


In [39]:
# Predicting using LogisticRegression for Breast Cancer
y_pred_lr_bc = clf_lr_bc.predict(x_test)

In [40]:
# Scoring LogisticRegression for Breast Cancer (Accuracy)
acc_lr_bc = accuracy_score(y_true, y_pred_lr_bc)
print("Accuracy Score LogisticRegression (Dataset Breast Cancer) =", acc_lr_bc)

Accuracy Score LogisticRegression (Dataset Breast Cancer) = 0.9473684210526315


In [41]:
# Scoring LogisticRegression for Breast Cancer (F1)
f1_lr_bc = f1_score(y_true, y_pred_lr_bc)
print("F1 Score LogisticRegression (Dataset Breast Cancer) =", f1_lr_bc)

F1 Score LogisticRegression (Dataset Breast Cancer) = 0.9558823529411765


### Play-Tennis Dataset

In [42]:
# LogisticRegression (Play Tennis)
clf_lr_pt = LogisticRegression().fit(x_train_pt, y_train_pt)
show_param(clf_lr_pt, "Parameters for LogisticRegression model Dataset Play Tennis:")

Parameters for LogisticRegression model Dataset Play Tennis:
C: 1.0
class_weight: None
dual: False
fit_intercept: True
intercept_scaling: 1
l1_ratio: None
max_iter: 100
multi_class: auto
n_jobs: None
penalty: l2
random_state: None
solver: lbfgs
tol: 0.0001
verbose: 0
warm_start: False


In [43]:
# Showing Coefficient and Bias from Logistic Regression in Decision Function for Play Tennis
print(clf_lr_pt.coef_,clf_lr_pt.intercept_)

[[-0.6364867   0.29289308  0.78462682  0.69810077]] [0.19971342]


In [44]:
# Predicting using LogisticRegression for Play Tennis
y_pred_lr_pt = clf_lr_pt.predict(x_test_pt)

In [45]:
# Scoring LogisticRegression for Play Tennis (Accuracy)
acc_lr_pt = accuracy_score(y_true_pt, y_pred_lr_pt)
print("Accuracy Score LogisticRegression (Dataset Play Tennis) =", acc_lr_pt)

Accuracy Score LogisticRegression (Dataset Play Tennis) = 0.6666666666666666


In [46]:
# Scoring LogisticRegression for Play Tennis (F1)
f1_lr_pt = f1_score(y_true_pt, y_pred_lr_pt)
print("F1 Score LogisticRegression (Dataset Play Tennis) =", f1_lr_pt)

F1 Score LogisticRegression (Dataset Play Tennis) = 0.8


## Soal 2e & 3e - Pembelajaran, Prediksi, dan Evaluasi dengan Algoritma Neural Network

### Breast Cancer Dataset

In [47]:
# Neural Network (Breast Cancer)
clf_mlp_bc = MLPClassifier(max_iter = 500).fit(x_train, y_train)
show_param(clf_mlp_bc, "Parameters for Neural Network model Dataset Breast Cancer:")

Parameters for Neural Network model Dataset Breast Cancer:
activation: relu
alpha: 0.0001
batch_size: auto
beta_1: 0.9
beta_2: 0.999
early_stopping: False
epsilon: 1e-08
hidden_layer_sizes: (100,)
learning_rate: constant
learning_rate_init: 0.001
max_fun: 15000
max_iter: 500
momentum: 0.9
n_iter_no_change: 10
nesterovs_momentum: True
power_t: 0.5
random_state: None
shuffle: True
solver: adam
tol: 0.0001
validation_fraction: 0.1
verbose: False
warm_start: False


In [48]:
# Showing weight matrix from the i-th element corresponding to layer i for Breast Cancer
clf_mlp_bc.coefs_

[array([[-5.76897167e-02, -1.48454643e-04,  1.55861789e-01, ...,
          3.00224953e-02, -6.57377728e-02, -1.15562187e-01],
        [-5.31686062e-02, -7.23703166e-02,  1.04793811e-01, ...,
         -1.57341174e-01,  5.56355058e-03, -1.99846013e-01],
        [-6.70292229e-04, -5.75154922e-04, -1.94347974e-04, ...,
         -1.28783231e-01,  8.94303644e-02, -1.14298180e-01],
        ...,
        [ 1.42338762e-03, -2.65405402e-02,  1.16626340e-02, ...,
          5.78299694e-02, -1.06863355e-01,  1.71721066e-01],
        [-6.05932686e-02,  8.13622704e-05, -1.91633650e-01, ...,
          7.59505755e-03,  7.41542964e-02, -1.30949453e-02],
        [-2.64266759e-05,  1.53071674e-03, -2.01804413e-01, ...,
          8.53053328e-02, -1.81981873e-01,  9.79393825e-02]]),
 array([[-2.36754888e-03],
        [-2.04251654e-04],
        [ 9.54793407e-02],
        [-2.08069403e-01],
        [-1.41431743e-01],
        [ 3.21332627e-02],
        [ 4.49175494e-02],
        [ 1.14983226e-01],
        [-7.3

In [49]:
# Predicting using Neural Network for Breast Cancer
y_pred_nn_bc = clf_mlp_bc.predict(x_test)

In [50]:
# Scoring Neural Network for Breast Cancer (Accuracy)
acc_nn_bc = accuracy_score(y_true, y_pred_nn_bc)
print("Accuracy Score Neural Network (Dataset Breast Cancer) =", acc_nn_bc)

Accuracy Score Neural Network (Dataset Breast Cancer) = 0.9122807017543859


In [51]:
# Scoring Neural Network for Breast Cancer (F1)
f1_nn_bc = f1_score(y_true, y_pred_nn_bc)
print("F1 Score Neural Network (Dataset Breast Cancer) =", f1_nn_bc)

F1 Score Neural Network (Dataset Breast Cancer) = 0.9275362318840579


### Play-Tennis Dataset

In [52]:
# Neural Network (Play Tennis)
clf_mlp_pt = MLPClassifier(max_iter = 1000).fit(x_train_pt, y_train_pt)
show_param(clf_mlp_pt, "Parameters for Neural Network model Dataset Play Tennis:")

Parameters for Neural Network model Dataset Play Tennis:
activation: relu
alpha: 0.0001
batch_size: auto
beta_1: 0.9
beta_2: 0.999
early_stopping: False
epsilon: 1e-08
hidden_layer_sizes: (100,)
learning_rate: constant
learning_rate_init: 0.001
max_fun: 15000
max_iter: 1000
momentum: 0.9
n_iter_no_change: 10
nesterovs_momentum: True
power_t: 0.5
random_state: None
shuffle: True
solver: adam
tol: 0.0001
validation_fraction: 0.1
verbose: False
warm_start: False


In [53]:
# Showing weight matrix from the i-th element corresponding to layer i for Play Tennis
clf_mlp_pt.coefs_

[array([[-4.02514227e-01,  8.94443004e-02, -2.26292316e-01,
          1.62237969e-01,  3.47406986e-01, -2.13034067e-01,
         -1.67629671e-01,  1.47268141e-01, -4.71776546e-02,
         -1.94032993e-01,  1.01975673e-02,  6.19194113e-13,
         -3.11379880e-01, -3.72616021e-01,  2.14932293e-01,
          2.82223749e-01, -1.72576989e-01,  8.62216234e-02,
         -1.76857816e-01, -1.58267000e-01,  2.56282155e-01,
          1.39919833e-01, -3.65266093e-02,  1.76086428e-02,
         -4.68671605e-01,  2.41129043e-01, -1.55701703e-01,
         -4.56570998e-01, -3.92927395e-01,  2.29130820e-10,
         -3.37902176e-02,  2.61350526e-01,  4.10385643e-01,
          2.46265363e-01,  1.02211560e-01, -1.58156654e-01,
         -8.96070344e-02,  3.84936881e-01, -1.07492691e-12,
          2.86158984e-01, -9.16846531e-02,  9.96739994e-03,
         -1.68186576e-01, -1.20624137e-01, -1.40228234e-14,
         -4.20800967e-01, -8.50256321e-02, -9.42712212e-02,
          1.97610664e-02,  2.33774400e-0

In [54]:
# Predicting using Neural Network for Play Tennis
y_pred_nn_pt = clf_mlp_pt.predict(x_test_pt)

In [55]:
# Scoring Neural Network for Play Tennis (Accuracy)
acc_nn_pt = accuracy_score(y_true_pt, y_pred_nn_pt)
print("Accuracy Score Neural Network (Dataset Play Tennis) =", acc_nn_pt)

Accuracy Score Neural Network (Dataset Play Tennis) = 1.0


In [56]:
# Scoring Neural Network for Play Tennis (F1)
f1_nn_pt = f1_score(y_true_pt, y_pred_nn_pt)
print("F1 Score Neural Network (Dataset Play Tennis) =", f1_nn_pt)

F1 Score Neural Network (Dataset Play Tennis) = 1.0


## Soal 2f & 3f - Pembelajaran, Prediksi, dan Evaluasi dengan Algoritma SVM

### Breast Cancer Dataset

In [57]:
# SVM (Breast Cancer)
clf_svm_bc = make_pipeline(StandardScaler(), SVC(gamma='auto')).fit(x_train, y_train)
show_param(clf_svm_bc, "Parameters for SVM model Dataset Breast Cancer:")

Parameters for SVM model Dataset Breast Cancer:
memory: None
steps: [('standardscaler', StandardScaler()), ('svc', SVC(gamma='auto'))]
verbose: False
standardscaler: StandardScaler()
svc: SVC(gamma='auto')
standardscaler__copy: True
standardscaler__with_mean: True
standardscaler__with_std: True
svc__C: 1.0
svc__break_ties: False
svc__cache_size: 200
svc__class_weight: None
svc__coef0: 0.0
svc__decision_function_shape: ovr
svc__degree: 3
svc__gamma: auto
svc__kernel: rbf
svc__max_iter: -1
svc__probability: False
svc__random_state: None
svc__shrinking: True
svc__tol: 0.001
svc__verbose: False


In [58]:
# SVM Support Vector [Showing SVM Model for Breast Cancer]
clf_svm_bc.named_steps["svc"].support_vectors_

array([[ 2.4702554 ,  0.10292725,  2.67043636, ...,  2.69558848,
         1.86311642,  0.71945803],
       [ 0.94590949, -0.6490507 ,  1.08421561, ...,  2.14760699,
         1.880914  ,  1.15107672],
       [ 0.09310778,  0.76521976,  0.12601758, ...,  0.33926808,
        -0.3324573 , -0.05540797],
       ...,
       [-0.14757842,  2.06680852, -0.21364055, ..., -0.8539616 ,
        -0.33892915, -0.7505242 ],
       [ 0.17036508,  0.02703957,  0.14670742, ..., -0.27355788,
        -0.69973456, -0.53390659],
       [-1.25562635, -0.89281114, -1.16666612, ...,  0.65740178,
         0.3309069 ,  2.22877631]])

In [59]:
# Predicting using SVM for Breast Cancer
y_pred_svm_bc = clf_svm_bc.predict(x_test)

In [60]:
# Scoring SVM for Breast Cancer (Accuracy)
acc_svm_bc = accuracy_score(y_true, y_pred_svm_bc)
print("Accuracy Score SVM (Dataset Breast Cancer) =", acc_svm_bc)

Accuracy Score SVM (Dataset Breast Cancer) = 0.9649122807017544


In [61]:
# Scoring SVM for Breast Cancer (F1)
f1_svm_bc = f1_score(y_true, y_pred_svm_bc)
print("F1 Score SVM (Dataset Breast Cancer) =", f1_nn_bc)

F1 Score SVM (Dataset Breast Cancer) = 0.9275362318840579


### Play-Tennis Dataset

In [62]:
# SVM (Play Tennis)
clf_svm_pt = make_pipeline(StandardScaler(), SVC(gamma='auto')).fit(x_train_pt, y_train_pt)
show_param(clf_svm_pt, "Parameters for SVM model Dataset Play Tennis:")

Parameters for SVM model Dataset Play Tennis:
memory: None
steps: [('standardscaler', StandardScaler()), ('svc', SVC(gamma='auto'))]
verbose: False
standardscaler: StandardScaler()
svc: SVC(gamma='auto')
standardscaler__copy: True
standardscaler__with_mean: True
standardscaler__with_std: True
svc__C: 1.0
svc__break_ties: False
svc__cache_size: 200
svc__class_weight: None
svc__coef0: 0.0
svc__decision_function_shape: ovr
svc__degree: 3
svc__gamma: auto
svc__kernel: rbf
svc__max_iter: -1
svc__probability: False
svc__random_state: None
svc__shrinking: True
svc__tol: 0.001
svc__verbose: False


In [63]:
# SVM Support Vector [Showing SVM Model for Breast Cancer]
clf_svm_bc.named_steps["svc"].support_vectors_

array([[ 2.4702554 ,  0.10292725,  2.67043636, ...,  2.69558848,
         1.86311642,  0.71945803],
       [ 0.94590949, -0.6490507 ,  1.08421561, ...,  2.14760699,
         1.880914  ,  1.15107672],
       [ 0.09310778,  0.76521976,  0.12601758, ...,  0.33926808,
        -0.3324573 , -0.05540797],
       ...,
       [-0.14757842,  2.06680852, -0.21364055, ..., -0.8539616 ,
        -0.33892915, -0.7505242 ],
       [ 0.17036508,  0.02703957,  0.14670742, ..., -0.27355788,
        -0.69973456, -0.53390659],
       [-1.25562635, -0.89281114, -1.16666612, ...,  0.65740178,
         0.3309069 ,  2.22877631]])

In [64]:
# Predicting using SVM for Play Tennis
y_pred_svm_pt = clf_svm_pt.predict(x_test_pt)

In [65]:
# Scoring SVM for Play Tennis (Accuracy)
acc_svm_pt = accuracy_score(y_true_pt, y_pred_svm_pt)
print("Accuracy Score SVM (Dataset Play Tennis) =", acc_svm_pt)

Accuracy Score SVM (Dataset Play Tennis) = 1.0


In [66]:
# Scoring SVM for Play Tennis (F1)
f1_svm_pt = f1_score(y_true_pt, y_pred_svm_pt)
print("F1 Score SVM (Dataset Play Tennis) =", f1_nn_pt)

F1 Score SVM (Dataset Play Tennis) = 1.0


## Soal 3 - Evaluasi seluruh hasil prediksi

In [67]:
# Showing all evaluation results [All Models]
headers = ["Accuracy_BC", "F1_BC", "Accuracy_PT", "F1_PT"]
indexes = ["DecisionTreeClassifier", "Id3Estimator", "K Means", "LogisticRegression", "Neural Network", "SVM"]
data = [
        [acc_dtc_bc, f1_dtc_bc, acc_dtc_pt, f1_dtc_pt],
        [acc_id3_bc, f1_id3_bc, acc_id3_pt, f1_id3_pt],
        [acc_km_bc, f1_km_bc, acc_km_pt, f1_km_pt],
        [acc_lr_bc, f1_lr_bc, acc_lr_pt, f1_lr_pt],
        [acc_nn_bc, f1_nn_bc, acc_nn_pt, f1_nn_pt],
        [acc_svm_bc, f1_svm_bc, acc_svm_pt, f1_svm_pt]
]
print("BC = Breast Cancer, PT = Play-Tennis")
print(tabulate.tabulate(data, headers, tablefmt="fancy_grid", showindex=indexes))

BC = Breast Cancer, PT = Play-Tennis
╒════════════════════════╤═══════════════╤══════════╤═══════════════╤══════════╕
│                        │   Accuracy_BC │    F1_BC │   Accuracy_PT │    F1_PT │
╞════════════════════════╪═══════════════╪══════════╪═══════════════╪══════════╡
│ DecisionTreeClassifier │      0.929825 │ 0.942029 │      0.666667 │ 0.666667 │
├────────────────────────┼───────────────┼──────────┼───────────────┼──────────┤
│ Id3Estimator           │      0.921053 │ 0.932331 │      0.666667 │ 0.666667 │
├────────────────────────┼───────────────┼──────────┼───────────────┼──────────┤
│ K Means                │      0.868421 │ 0.899329 │      1        │ 1        │
├────────────────────────┼───────────────┼──────────┼───────────────┼──────────┤
│ LogisticRegression     │      0.947368 │ 0.955882 │      0.666667 │ 0.8      │
├────────────────────────┼───────────────┼──────────┼───────────────┼──────────┤
│ Neural Network         │      0.912281 │ 0.927536 │      1        │ 1 

## Soal 4 - Analisis hasil Accuracy dan F1 Seluruh Algoritma Pembelajaran

*Accuracy Score* adalah suatu metode penilaian yang menghitung seberapa besar tingkat akurasi subset dari model dalam klasifikasi multi label. Selain itu, metode ini juga menghitung tingkat akurasi dalam memprediksi dengan menjalankan data testing. Caranya adalah dengan menilai seberapa besar persentase kebenaran yang diprediksi oleh model dalam dataset untuk evaluasi.

Metode penilaian lainnya adalah *F1 Score* yang merupakan fungsi dari nilai *precision* dan *recall*. *Precision* berhubungan dengan seberapa tepat model memprediksi positif dari seluruh data yang diberikan label positif oleh model, sedangkan *recall* berhubungan dengan banyaknya data yang benar diprediksi positif dari seluruh data yang memang terkategori positif. Sehingga, *F1 Score* merupakan fungsi yang mempertimbangkan *precision* dan *recall* sehingga lebih baik digunakan ketika terdapat cost lebih ketika memprediksi *False Negative* dan *False Positive*.

Untuk dataset *Breast Cancer*, nilai yang paling baik digunakan sebagai metrik untuk menilai model adalah *F1 Score* karena mempertimbangkan *Precision* dan *recall*. Sehingga sebisa mungkin menghindari *False Negative* atau *False Positive*. Untuk dataset Play Tennis, berbagai *random* *state* biasanya memiliki *F1 Score* dan *Accuracy Score* yang sama. Namun, banyak terdapat *random* *state* yang memiliki nilai *F1 Score* yang lebih baik dibandingkan dengan *Accuracy Score*.

Dataset Play Tennis memiliki lima belas baris data yang delapan puluh persennya digunakan untuk *training* model dan dua puluh persennya digunakan untuk set evaluasi. Status pembagian dataset tersebut ditandai oleh suatu *state* *random*. Saat dilakukan *training* model dengan suatu *state* *random*, diperoleh berbagai hasil penilaian yang seragam saat dilakukan *scoring* menggunakan metode *Accuracy Score* dan *F1 Score*. Namun demikian, hal tersebut tidak berlaku saat dilakukan pembentukan *training* model lain dengan *state* *random* yang lain juga. Angka hasil *scoring* saat dilakukan perubahan *state* *random* yang berbeda tersebut memiliki margin perubahan yang sangat besar. Melalui scoring dengan metode *Accuracy Score*, pemberlakuan *evaluation set* sebesar dua puluh persen dari lima belas baris data yang berarti penilaian dilakukan dengan tiga set dapat menghasilkan lebih dari 2 hasil yang berbeda. Terdapat kasus-kasus dimana penilaian model menggunakan *Accuracy Score* menghasilkan angka 0.33, 0.66, dan juga 1. Tentu hasil ini membuat *training* model kurang akurat dan sangat bergantung dengan *state* *random* di bagian awal saat pembagian dataset. Hal ini dapat disebabkan oleh jumlah data yang sangat sedikit dengan parameter atribut yang dapat dibilang tidak sedikit atau simpel. Oleh karena itu, perbedaan pembagian dataset untuk *training* dan evaluasi dapat memberikan hasil penilaian yang sangat berbeda. Tanpa patokan data yang beragam untuk *training* model, prediksi yang dilakukan pada *evaluation set* oleh model dapat mengalami banyak kegagalan karena kurang relevannya model hasil *training* dengan *evaluation set*.

Sedangkan, dataset *Breast Cancer* hanya sedikit terpengaruhi *state* awal yang *random*. Hal ini dikarenakan dataset yang memiliki jumlah data yang banyak dan variasi yang baik, sehingga hasil yang didapatkan dengan *random* *state* yang berbeda lebih konsisten dengan margin yang sedikit. Ketika *random* *state* pemisahan data berganti, hasil *Accuracy Score* dan *F1 Score* dari tiap model yang melakukan *training* dengan dataset *Breast Cancer* tidak mengalami perubahan yang signifikan.

Dari seluruh algoritma yang ada, algoritma yang menghasilkan model dengan hasil fungsi prediksi paling baik untuk *Breast Cancer* secara konsisten dengan berbagai *random* *state* adalah SVM atau *Support Vector Machines*. Sedangkan, algoritma yang dengan konsisten menghasilkan nilai prediksi yang paling buruk diantara model lainnya adalah K Means.